pylearn2/sandbox/nlp/models/mlp.py

```diff
@@ -189,3 +198,303 @@ def get_l1_weight_decay(self, coeff):
         assert isinstance(coeff, float) or hasattr(coeff, 'dtype')
         W, = self.transformer.get_params()
         return coeff * abs(W).sum()
+
+
+class ContrastiveProbabilityLayer(Layer):
+    """
+    Predicts the probability of a sample to come from a positive distribution
+    or a negative distribution. For each sample (i.e. a row in the input),
+    this layer produces *two* sets of probabilities. The first set has n
+    probabilities which says "if the sample is of class i then the probability
+    it comes from the positive distribution is p_i". The second set is the
+    complement of the first.
```

Comment:
It's ugly, I know. The point was to let users provide their own set of positive and negative examples. The first set of output is compared to positive examples while the second set of output negative examples.